google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -427,61 +434,166 @@ int computeDeadlineSeconds() {
   @InternalApi
   void extendDeadlines() {
     int extendSeconds = getMessageDeadlineSeconds();
-    List<PendingModifyAckDeadline> modacks = new ArrayList<>();
-    PendingModifyAckDeadline modack = new PendingModifyAckDeadline(extendSeconds);
+    int numAckIdToSend = 0;
+    Map<Integer, ModackWithMessageFuture> modackWithMessageFutureByExtensionTimeMap =
+        new HashMap<Integer, ModackWithMessageFuture>();
     Instant now = now();
     Instant extendTo = now.plusSeconds(extendSeconds);
 
     for (Map.Entry<String, AckHandler> entry : pendingMessages.entrySet()) {
       String ackId = entry.getKey();
       Instant totalExpiration = entry.getValue().totalExpiration;
       if (totalExpiration.isAfter(extendTo)) {
-        modack.ackIds.add(ackId);
+        ModackWithMessageFuture modackWithMessageFuture =
+            modackWithMessageFutureByExtensionTimeMap.getOrDefault(
```

Comment:
You can just use computeIfAbsent here, instead of getOrDefault. That way, you can get rid of the step below where you put the modackWithMessageFuture into the map